### PR TITLE
OSSA ownership optimization RAUW utility fixes.

### PR DESCRIPTION
**Pull request by atrick**
Verify that the OwnershipRAUWUtility always preserves the original
borrow scope by exhaustively switching over OperandOwnership.

And related cleanup.

### CODE REVIEW

**Comment by atrick** [comment]
@gottesmm just need to get this cleanup and self-checking in place to I can proceed to add functionality. I haven't self-reviewed or tested it yet

**Comment by atrick** [comment]
@swift-ci test

File: include/swift/SIL/OwnershipUtils.h [hunk 1]
```diff
@@ -76,12 +76,14 @@ inline bool isForwardingConsume(SILValue value) {
...
+public:
+  static ForwardingOperand get(Operand *use) { return ForwardingOperand(use); }
 
-  ForwardingOperand(Operand *use) : use(use) {}
+private:
+  Operand *use = nullptr;
 
 public:
-  static ForwardingOperand get(Operand *use);
+  ForwardingOperand(Operand *use);
```
Comment (atrick) [inline include/swift/SIL/OwnershipUtils.h:86]: I'll probably make this construct "explicit" in the next PR.

**Comment by atrick** [comment]
@swift-ci smoke test

File: include/swift/SIL/OwnershipUtils.h [hunk 2]
```diff
@@ -665,25 +663,26 @@ struct InteriorPointerOperand {
...
   /// The algorithm that is used to determine what the verifier will consider to
-  /// be implicit uses of the given address. Used to implement \see
-  /// getImplicitUses.
+  /// be transitive uses of the given address. Used to implement \see
+  /// findTransitiveUses.
   static bool
-  getImplicitUsesForAddress(SILValue address,
-                            SmallVectorImpl<Operand *> &foundUses,
-                            std::function<void(Operand *)> *onError = nullptr);
+  findTransitiveUsesForAddress(SILValue address,
```
Comment (gottesmm) [inline include/swift/SIL/OwnershipUtils.h:683]: Eventually we are going to want to represent convert_escape_to_no_escape as an interior pointer and its result is not an address. I am not sure what the appropriate name is though. Maybe findTransitiveProtectedUses? With the idea that we are finding the uses that the borrow scope is ensuring are valid... open to better ideas.

File: lib/SILOptimizer/Utils/OwnershipOptUtils.cpp [hunk 3]
```diff
@@ -99,70 +99,88 @@ insertOwnedBaseValueAlongBranchEdge(BranchInst *bi, SILValue innerCopy,
...
+    case OperandOwnership::TrivialUse:
+    case OperandOwnership::ForwardingConsume:
+    case OperandOwnership::DestroyingConsume:
+      llvm_unreachable("this operand cannot handle a guaranteed use");
+
+    case OperandOwnership::ForwardingUnowned:
+    case OperandOwnership::PointerEscape:
+      return false;
+
+    case OperandOwnership::InstantaneousUse:
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/OwnershipOptUtils.cpp:128]: Thanks for updating this to use OperandOwnership! This code was written before that was created IIRC.

File: lib/SILOptimizer/Utils/OwnershipOptUtils.cpp [hunk 4]
```diff
@@ -193,23 +223,21 @@ static bool canFixUpOwnershipForRAUW(SILValue oldValue, SILValue newValue) {
...
-  if (oldOwnershipKind != OwnershipKind::None)
-    return true;
+  return true;
+}
 
-  // Ok, we have an old ownership kind that is OwnershipKind::None and a new
-  // ownership kind that is not OwnershipKind::None. In that case, for now, do
-  // not perform this transform.
-  return false;
+static bool canFixUpOwnershipForRAUW(SILValue oldValue, SILValue newValue,
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/OwnershipOptUtils.cpp:229]: Can you add a comment here similar to the comment on hasValidRAUWOwnership explaining that you also check for transitive uses?

File: include/swift/SIL/OwnershipUtils.h (hunk 2 above)
Comment (gottesmm) [reply include/swift/SIL/OwnershipUtils.h:683]: Now that I am reading in a little more detail, maybe we would create a separate findTransitiveUsesForObject for the convert_escape_to_no_escape case.

**Review by gottesmm — APPROVED** [review]
Some small comments/thoughts. Overall looks good!

File: lib/SILOptimizer/Utils/OwnershipOptUtils.cpp (hunk 4 above)
Comment (atrick) [reply lib/SILOptimizer/Utils/OwnershipOptUtils.cpp:234]: oops that comment got dropped. fixed it

File: include/swift/SIL/OwnershipUtils.h (hunk 2 above)
Comment (atrick) [reply include/swift/SIL/OwnershipUtils.h:683]: I wanted to revisit these helpers but didn't want to put the time in on it for this PR

**Comment by atrick** [comment]
@swift-ci smoke test and merge

**Pull request merged**